src/function/gds/output_writer.cpp

```diff
@@ -83,31 +81,10 @@ void PathsOutputWriter::write(processor::FactorizedTable& fTable, nodeID_t dstNo
     while (!curPath.empty()) {
         auto top = curPath[curPath.size() - 1];
         auto topNodeID = top->getNodeID();
-        if (topNodeID == sourceNodeID) {
+        if (topNodeID == sourceNodeID && top->getIter() == 1) {
             // check path
-            if (checkPathNode(curPath)) {
-                beginWritingNewPath(curPath.size());
-                if (extendFromSource) {
-                    // Write path in reverse direction because we append ParentList from dst to src.
-                    for (auto i = 1u; i < curPath.size(); i++) {
-                        auto p = curPath[curPath.size() - 1 - i];
-                        addNode(p->getNodeID(), i - 1);
-                        addEdge(p->getEdgeID(), p->isFwdEdge(), i);
-                    }
-                    auto lastPathElement = curPath[curPath.size() - 1];
-                    addEdge(lastPathElement->getEdgeID(), lastPathElement->isFwdEdge(), 0);
-                } else {
-                    // Write path in original direction because computation started from dst node.
-                    // We want to present result in src->dst order.
-                    for (auto i = 0u; i < curPath.size() - 1; i++) {
-                        auto p = curPath[i];
-                        addNode(p->getNodeID(), i);
-                        addEdge(p->getEdgeID(), p->isFwdEdge(), i);
-                    }
-                    auto lastPathElement = curPath[curPath.size() - 1];
-                    addEdge(lastPathElement->getEdgeID(), lastPathElement->isFwdEdge(),
-                        curPath.size() - 1);
-                }
+            if (checkPathNodeMask(curPath) && checkSemantic(curPath)) {
```

Comment:
My comment is the same. Doing these check after each path enumeration can be arbitrarily slower than doing it as you do the DFS-based path enumeration. So similar to my comment in the other PR, I would do this as part of the DFS-based enumeration. The most obvious way is to have some auxiliary c++ map that you keep each time addEdge is called if the semantics is trail or acyclic and call checkSemantic before adding a new edge. The logic around where you make the addEdge calls might need to change somewhat as well.

Reply:
As we discussed, we will do this optimization together with node filter optimization.